### PR TITLE
feat(process): split crawling and grouping

### PR DESCRIPTION
**Pull request by BioPhoton**

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -18,14 +18,20 @@ import {
...
 
-  const processors = [addGrouping, addUniqueKey];
-  const processedDeprecations = (await processors.reduce(
+  // Crawling Phase
+  const crawlingPhaseProcessors = [addUniqueKey, addCommentToRepository];
+  const processedCrawledDeprecations = (await crawlingPhaseProcessors.reduce(
     async (deps, processor) => await processor(config, await deps),
     Promise.resolve(deprecations)
   )) as Deprecation[];
+  await generateRawJson(config, processedCrawledDeprecations, { tagDate });
```
Comment (timdeschryver) [inline lib/index.ts:27]: This means that the group won't be included in the JSON output.
Is this OK?

Should we resave the JSON after the grouping phase?

**Comment by BioPhoton** [comment]
Yes, we should save it.

Also this is not a complete pr as we miss the question if we want to start grouping or not.

Here I also would need feedback.

**Comment by timdeschryver** [comment]
How can I help?

**Comment by BioPhoton** [comment]
I just started to read the code  in more detail.
I would unify the processes to always return `Promise<Deprecation[]>` so we can better chain them.

For adding uuids and grouping I will update the methods to either look for existing output to update or create new one.
This works easy for JSON. so I will start with that and the "rendering" to the various formats markdown, json etc is done as last step. This step, for now, will always overwrite the generated output.

This has following implications:

Maintaining rawDeprecation data and enriching it could be 2 separate processes I suggest to have a JSON data structure that maintains the raw data, UUIDs and groups. This structure can be updated programmatically and does not contain and custom information. 

Downside, we have a data structure in addition to the markdown/json output.

**Comment by BioPhoton** [comment]
Please check my latest updates.

I implemented the updating of deprecations. Not perfect, but imho good enough for now.

**Comment by BioPhoton** [comment]
Merged conflicts

File: lib/output-formatters/generate-raw-json.ts [hunk 2]
```diff
@@ -10,20 +10,48 @@ export async function generateRawJson(
...
+    };
+  }
+
+  const json = JSON.stringify(content, null, 4);
   writeFileSync(join(config.outputDirectory, `${config.gitTag}.json`), json);
 }
+
+function upsetrDeprecations(oldDeprecations: Deprecation[], newDeprecations: Deprecation[]): Deprecation[] {
+  const upsertedDeprecations = [...oldDeprecations];
+  upsertedDeprecations.forEach(d => upsetrDeprecation(upsertedDeprecations, d));
```
Comment (timdeschryver) [inline lib/output-formatters/generate-raw-json.ts:49]: Doesn't this have to be the new deprecations, in order to add them?

```suggestion
  newDeprecations.forEach(d => upsetrDeprecation(upsertedDeprecations, d));
```

File: lib/output-formatters/generate-raw-json.ts [hunk 3]
```diff
@@ -10,20 +10,48 @@ export async function generateRawJson(
...
+      date: options.tagDate,
+      deprecations: rawDeprecations
+    };
+  }
+
+  const json = JSON.stringify(content, null, 4);
   writeFileSync(join(config.outputDirectory, `${config.gitTag}.json`), json);
 }
+
+function upsetrDeprecations(oldDeprecations: Deprecation[], newDeprecations: Deprecation[]): Deprecation[] {
```
Comment (timdeschryver) [inline lib/output-formatters/generate-raw-json.ts:47]:
```suggestion
function upsertDeprecations(oldDeprecations: Deprecation[], newDeprecations: Deprecation[]): Deprecation[] {
```

File: lib/output-formatters/generate-raw-json.ts [hunk 4]
```diff
@@ -10,20 +10,48 @@ export async function generateRawJson(
...
+  const upsertedDeprecations = [...oldDeprecations];
+  upsertedDeprecations.forEach(d => upsetrDeprecation(upsertedDeprecations, d));
+  return upsertedDeprecations;
+}
+
+function upsetrDeprecation(oldDeprecations: Deprecation[], deprecation: Deprecation) {
```
Comment (timdeschryver) [inline lib/output-formatters/generate-raw-json.ts:53]:
```suggestion
function upsertDeprecation(oldDeprecations: Deprecation[], deprecation: Deprecation) {
```

**Comment by timdeschryver** [comment]
> I would unify the processes to always return Promise<Deprecation[]> so we can better chain them.

👍 - I didn't do it for the output formatters because they shouldn't update the deprecations.  

> Maintaining rawDeprecation data and enriching it could be 2 separate processes I suggest to have a JSON data structure that maintains the raw data, UUIDs and groups. This structure can be updated programmatically and does not contain and custom information.

I like that idea. We should try it out 🙂
This would also mean you would be able to "restart" the grouping. I like it.
We just have to make sure the UUIDs are in the raw output (because we use them to add comments to the repo).

💡 The grouped output has to be in a seperate file (as you mentioned), what about the following structure?
Perhaps this is what you had in mind, because at first I thought it would be the same structure but with the added group.

```
{
  'uuid1': 'group1',
  'uuid2': 'group2',
  'uuid3': 'group1',
}
```  

or perhaps even better as it would be more flexible if we would want to have a deprecation in multiple groups.

```
{
  'group1': ['uuid1', 'uuid2'],
  'group2': ['uuid3'],
}
```

File: lib/output-formatters/generate-raw-json.ts (hunk 2 above)
Comment (timdeschryver) [reply lib/output-formatters/generate-raw-json.ts:49]: Also, it doesn't have to be done in this PR but we should also remove deprecations?
For example, deprecation is added, output is generated, deprecation is removed, output is generated and the deprecation is removed in the output.

The current implementation would update deprecations and add new deprecations, right?

File: lib/index.ts (hunk 1 above)
Comment (BioPhoton) [reply lib/index.ts:27]: Done

File: lib/output-formatters/generate-raw-json.ts (hunk 2 above)
Comment (BioPhoton) [reply lib/output-formatters/generate-raw-json.ts:49]: Right! Will fix it!

**Comment by BioPhoton** [comment]
I'll move the grouping discussion into another issue.

What about the process? Is the current version good enough for now :)?

**Pull request merged**